generator/generator/generator.go

```diff
@@ -70,6 +71,38 @@ func Generate(dir string, fileName string, wasm []byte, initFunction string) err
 		return err
 	}
 
+	compiledBuilder, err := builder.Compile(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Generate any missing imports.
+	exportedFunctions := compiledBuilder.ExportedFunctions()
+	importedFunctions := compiledModule.ImportedFunctions()
+	missingFunctions := map[string]api.FunctionDefinition{}
+	for i := range importedFunctions {
+		module, importName, _ := importedFunctions[i].Import()
+		if module == "env" {
```

Comment:
@Danlock I thought host function calls would always end up in the env module, or am I wrong?

Reply:
env is just the default name, it can be changed.
https://clang.llvm.org/docs/AttributeReference.html#import-module

Also I imagine the same problem could occur if someone imported wasip2 or 3 functionality, since the generator doesn't import those.

Reply:
@Danlock Did not know about that! :) I have just implemented a change that should work with any module name.

Reply:
@Danlock I looked at your comments in your repository:

```
	/*
		Unable to figure out how to register certain types with embind...
		Passing a []byte to C++ expecting a std:vector<unsigned char> results in Cannot call OCREngine.loadRawImage due to unbound types: NSt3__26vectorIhNS_9allocatorIhEEEE
		Passing an int to C++ expecting a const unsigned char * results in a Cannot call OCREngine.loadRawImage due to unbound types: PKh
		Passing an int to C++ expecting a void * results in a Cannot call OCREngine.loadRawImage due to unbound types: Pv
		Passing any kind of raw pointer seems to result in a variety of unbound types errors.
		Passing a string seems to work, although that will incur a Go copy of []byte to string, and then additional embind copies to the WASM memory.
		Instead we just skip over embind altogether and directly call pixReadMem on our previously written bytes. Seems to be as efficient as we can get.
	*/
```

You probably want to look at memory views, that will give you a direct view of the WebAssembly memory inside Go, that will give you the data without any copying.

Reply:
typed_memory_view was the reason I started using this library. I wasn't able to get it working though, and I ran into trouble passing raw pointers as well. I looked inside your tests and couldn't find an example that was passing a []byte to C++ using memory view, just vectors into Go.

But that was my first commit to get this working and able to be tested. I was gonna try memory view's again to decrease copying once I got some tests up. Do you think I will be able to directly pass a []byte into C++ using emscripten::val as described in this issue?

https://github.com/emscripten-core/emscripten/issues/5519

Reply:
@Danlock I'm playing with it a bit now, and got the model loading to work with that. What the easiest way is to make a container class in C++:

```
class Model {
public:
    Model(int size) {
        size_ = size;
        bytes_ = (const char *) malloc(size);
    }

    ~Model() { free((void *) bytes_); }

    emscripten::val Data() const {
      return emscripten::val(emscripten::typed_memory_view(size_, (unsigned char *) bytes_));
    }

    const char * Bytes() const { return bytes_; }

    size_t Size() const { return size_; }

private:
    size_t size_;
    const char *bytes_;
};
```

You can then expose it like this:
```
  class_<Model>("Model").constructor<int>().function("data", &Model::Data);
```

In Go you can then fill up the data like this:
```
	ocrModel, err := gen.NewClassModel(embEng, ctx, int32(len(langModel)))
	if err != nil {
		return "", fmt.Errorf(logPrefix+" gen.NewClassModel %w", err)
	}

	modelData, err := ocrModel.Data(ctx)
	if err != nil {
		return "", fmt.Errorf(logPrefix+" Model.Data %w", err)
	}

	// Copy the model data into WebAssembly memory.
	copy(modelData.([]uint8), langModel)
```

Now looking into how to do this for the model as well.

Reply:
Thanks so much for your help. I managed to get it working using your examples. 

However I am still curious about some other things. 

Why can't a raw pointer like const char * be used as  a function parameter without errors like "Cannot call OCREngine.loadRawImage due to unbound types: PKh"? Is there some way to register those pointers that I am unaware of?

Reply:
I don't know, I think it's just not supported right now to pass arbitrary pointers as arguments through Embind, maybe you can pass them by wrapping them with emval? The only pointers that are supported right now seem to be class pointers and `std::unique_ptr`.

Reply:
OK. I was passing them anyway with a size_t in the C++ function parameters that would get casted to a const char *. Grimy, but works. Maybe that should be documented somewhere as a workaround?